### ISSUE TITLE
How to get picking listener direct from Scene Object  

### ISSUE THREAD

**Issue by DanielNogueiraAndroid** [Samsung/GearVRf#122]
There is a way to get direct from the GVRSceneObject if he is picked? like:
```java
   GVRSceneObject object;
   // ...
      object.setPickingListener(new PickingListener(){
       public void onPick(GVREyePointeeHolder e) {
       }
    });
```
or 
```java
public class Mic extends GVRSceneObject{
//...
    @Override
    public void onPick(GVREyePointeeHolder e) {
    }

}
```

**Comment by JonShemitz** [comment]
No, there is not. Nor am I sure that this is desirable. One doesn't always care about only what is exactly in  the center of the field of view: Imagine, for example, an app that tracks the position of the user's hands, and wants to use picking to figure out what she might be touching.

**Comment by DanielNogueiraAndroid** [comment]
Ok so the method can call onCenter(GVREyePointeeHolder e) instead onPick(GVREyePointeeHolder e)?

**Comment by DanielNogueiraAndroid** [comment]
By this moment I'm controlling 50 objects (and rising) "onCenter" inside a GVRScript for ... this is not looking good.

**Comment by thomasflynn** [comment]
This kind of picking is not an unusual use case, but it does require deeper thought than just onPick() (which i assume would map to intersection with object and tapping the touchpad?).  Just off-hand, I can think of onEnter, onLeave, onSelected, onDeselected, and you may want to apply a filter based on the type of object.  This type of thing has been thought out before in APIs like Inventor and X3D.  We will need to give the same kind of thought here as well.

**Comment by JonShemitz** [comment]
In addition to the `onPick()` _vs._ `onCenter()` issue (and the Other APIs issue that @thomasflynn  cites) I'm having some trouble seeing why this is desirable.

First, the current `pickScene()` and `findObjects()` APIs sort hits by their distance from the camera. This lets you do things like pay attention to transparency and walk the list until you get to the object the user 'means to' interact with. Or model translucency, and handle hits until the 'beam' is totally attenuated.

Second, I am sure that it's cleaner to just add `GVRSceneObject` descendants to a `GVRScene` and let them Do The Right Thing when picked than to walk the list and Do The Right Thing in the loop ... but, honestly, the `instanceof` and cast overhead in a loop like
```java
    for (GVRPickedObject picked: findObjectsResult) {
      GVRSceneObject object = picked.getHitObject();
      if (object instanceof MyAppObject) {
        ((MyAppObject) object).myOnPick();
        }
      }
```
is really quite minimal compared to the cost of picking itself, which has to do non-trivial math on every pickable object.

**Comment by DanielNogueiraAndroid** [comment]
If we avoid using intanceof, and instead get this info directly on GVRSceneObject, we'll make the bottleneck, that is handling every state of eyepicking, smaller.

For instance, we have to control elements when they are selected and when they are not. If this control was held to GVRSceneObject, we wouldn't have to do it checking the eyepointeeholder.
```java
if (eyePointeeHolders.length == 0) {
            for (GVRSceneObject object : mGVRContext.getMainScene().getWholeSceneObjects()) {
                if (object instanceof SphereFlag && object.equals(lastItem)) {
                    ((SphereFlag) object).restoreObject(); 
                }
             }
} else {
            for (GVREyePointeeHolder eph : eyePointeeHolders) {
                for (GVRSceneObject object : mGVRContext.getMainScene().getWholeSceneObjects()) {
                    if (eph.getOwnerObject().equals(object) && object instanceof SphereFlag) {
                        ((SphereFlag) object).doSomething();
                    }
                }
             }

```

**Comment by JonShemitz** [comment]
```java
for (GVRSceneObject object : mGVRContext.getMainScene().getWholeSceneObjects()) {
   if (object instanceof SphereFlag && object.equals(lastItem)) {
     ((SphereFlag) object).restoreObject(); 
```

This seems like a lot of overhead, and I'm not sure it's any different that 

    if (lastItem != null) {
      lastItem.restorteObject();
    }
____

Similarly, it's not all obvious why the second branch enumerates all the eye pointee holders in an outer loop, then enumerates all the objects in the scene graph in an inner loop. Picking should never return an object that's not in the scene graph.
